### ISSUE TITLE
Time component to map matching?

### ISSUE THREAD

**Issue by durkie** [valhalla/valhalla#1495]
Hi Valhalla folks -- thanks for the great project. I'm using the map matching service and for my application it's important that the matched polyline closely approximate the original polyline in terms of overall length. So I'm wondering if it's possible to make the map matching algorithm less generous with the paths it generates.

Exhibit A:
GPS trace:
![image](https://user-images.githubusercontent.com/3200608/44828959-48fb6f00-abe8-11e8-97ef-062d6ac5641e.png)

matched polyline:
![image](https://user-images.githubusercontent.com/3200608/44828980-629cb680-abe8-11e8-8576-336149cf085d.png)

So the user turned left off of Monroe, turned right on to Golf Circle, noticed a (off-map) cut-through alley, and made a U-turn so that they could travel that way instead. The matched polyline just has them traveling the entirety of Golf Circle.

I understand why the matching algorithm came up with this: it sees two GPS points on each side of Golf Circle, several points in the middle of Golf Circle that are unmatched and ignored, and it deduces that you must have traveled the entire length of Golf Circle. The problem is that this didn't happen and the matched route is a fair bit longer than the original, especially when these errors accumulate.

Is there a way to get a matched polyline that more closely resembles the original route? I feel like the hidden markov model is ignoring a time component of the polyline in assigning its probabilities: if you assume all the gps points are equally spaced in time (or were able to provide this information), then traveling the route suggested by the map matching algorithm would require traveling at superhuman/illegal speeds.

Exhibit B:
GPS trace:
![image](https://user-images.githubusercontent.com/3200608/44829585-89102100-abeb-11e8-92b6-138c59308694.png)

matched polyline:
![image](https://user-images.githubusercontent.com/3200608/44829614-ab09a380-abeb-11e8-98f5-e9f910439ef2.png)

Similar situation here: the user did a U-turn on 26th St, but did not travel all the way to the end of it before doing so. 

It seems like if there were a way to provide something like a `gps_interval` parameter to the map matcher then it might help inform some of the choices it makes and could potentially tie in with the costing mode. Like if you were using `pedestrian` costing and a 1 Hz GPS interval, then you'd expect each successive point in the trace to change in distance by 1-4 m and could perhaps be more skeptical about points that varied more than this.

**Comment by kevinkreiser** [comment]
@durkie yes, in short it is possible to make the results less generous by supplying the time perameter for each location. have a look a the map matching test, you'll see how the requests formulated there specify the `time` field on each gps location: 

https://github.com/valhalla/valhalla/blob/master/test/mapmatch.cc#L61

basically its just epoch seconds. appologies that its not been documented yet!

where the time rules out a path which is impossibly long (there is a little slack though) you may have discontinuities in your trace, so please be aware of that. these will be annotated so they are easy to see.

**Comment by durkie** [comment]
Ah cool -- so I could use this right now then? For instance in the `trace_attributes` [examples](https://github.com/valhalla/valhalla-docs/blob/master/map-matching/api-reference.md#example-trace_attributes-requests), I could specify something like:
```
{"shape":[{"lat":39.983841,"lon":-76.735741,"time":12345},{"lat":39.983704,"lon":-76.735298,"time:12346}...
```

Does that seem right?

**Comment by kevinkreiser** [comment]
@durkie Thats exactly right yes. Sorry we are a bit behind on documentation!

**Comment by durkie** [comment]
This isn't quite having the effect I'd like yet. If I input the points from the above polyline with a 1 second interval between each point, it does avoid traveling the entire length of Golf Circle but also incorrectly matches many other areas. If I switch to a 2 second interval then it erroneously resumes traveling the entirety of Golf Circle, even though this would still imply traveling at very fast speeds.

Any other levers I could pull here to affect the matching behavior around measurement time / implied speed? I have tried a few different values for `breakage_distance`, `interpolation_distance`, and `max_route_time_factor`, but haven't found a clear direction to go with those.
